### PR TITLE
Histogram Optimized Tree Grower

### PR DESCRIPTION
**Pull request by hcho3**
* Add a brand new updater to support histogram-based algorithm, which buckets continuous features into discrete bins to speed up training. To use it, set `tree_method=hist` to configuration.
* Support multiple tree growing strategies. For now, two policies are supported:
  * `grow_policy=depthwise` (default):  favor splitting at nodes closest to the root, i.e. grow depth-wise.
  * `grow_policy=lossguide`: favor splitting at nodes with highest loss change
* Improve single-threaded performance
  * Unroll critical loops
  * Introduce specialized code for dense data (i.e. no missing values)
* Additional training parameters: `max_leaves`, `max_bin`, `grow_policy`, `verbose`

### CODE REVIEW

**Comment by tqchen** [comment]
please elaborate a bit on how each additional parameter can be used (max_leaves, max_bins, grow_policy)

**Comment by hcho3** [comment]
@tqchen Should I add a separate `README` for the new parameters?

**Comment by tqchen** [comment]
no, just paste the comments in the PR to give it more context to others

**Comment by hcho3** [comment]
Working on fixes to pass CI...

**Comment by hcho3** [comment]
CI should pass now...

**Comment by tqchen** [comment]
please fix my last comment, go through the code again and I will merge this PR in. Thanks!

**Comment by hcho3** [comment]
Opened an issue to document features: #1950

**Comment by tqchen** [comment]
To make the feature available in R, add the file to 

https://github.com/dmlc/xgboost/blob/master/amalgamation/xgboost-all0.cc

File: src/common/row_set.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                       unsigned left_node_id,
+                       unsigned right_node_id) {
+    const Elem e = elem_of_each_node_[node_id];
+    const unsigned nthread = row_split_tloc.size();
+    CHECK(e.begin != nullptr);
+    bst_uint* all_begin = dmlc::BeginPtr(row_indices_);
+    bst_uint* begin = all_begin + (e.begin - all_begin);
+
+    bst_uint* it = begin;
+    for (bst_omp_uint tid = 0; tid < nthread; ++tid) {
```
Comment (tqchen) [inline src/common/row_set.h]: add a comment: can be parallelized

File: src/tree/updater_fast_hist.cc [hunk 2]
```diff
@@ -0,0 +1,702 @@
...
+    TStats stats;
+    /*! \brief loss of this node, without split */
+    bst_float root_gain;
+    /*! \brief weight calculated related to current data */
+    float weight;
+    /*! \brief current best solution */
+    SplitEntry best;
+    // constructor
+    explicit NodeEntry(const TrainParam& param)
+        : stats(param), root_gain(0.0f), weight(0.0f){
```
Comment (tqchen) [inline src/tree/updater_fast_hist.cc]: space between ```){```

File: src/tree/updater_fast_hist.cc [hunk 3]
```diff
@@ -0,0 +1,702 @@
...
+          qexpand_->push(ExpandEntry(cright, p_tree->GetDepth(cright),
+                                     snode[cright].best.loss_chg));
+
+          ++num_leaves;  // give two and take one, as parent is no longer a leaf
+        }
+      }
+
+      // set all the rest expanding nodes to leaf
+      while (!qexpand_->empty()) {
+        const int nid = qexpand_->top().nid;
```
Comment (tqchen) [inline src/tree/updater_fast_hist.cc]: add comment:
This post condition is not needed in current code, but may be necessary when there are stopping rule that leaves qexpand non-empty

File: src/tree/updater_fast_hist.cc [hunk 4]
```diff
@@ -0,0 +1,702 @@
...
+                        DMatrix* p_fmat,
+                        RegTree* p_tree) {
+      double gstart = dmlc::GetTime();
+
+      std::vector<int> feat_set(p_fmat->info().num_col);
+      std::iota(feat_set.begin(), feat_set.end(), 0);
+      int num_leaves = 0;
+
+      double tstart;
+      double init_data = 0;
```
Comment (tqchen) [inline src/tree/updater_fast_hist.cc]: variable name better inform what it is, change to ```time_init_data```, etc

File: src/tree/updater_fast_hist.cc [hunk 5]
```diff
@@ -0,0 +1,702 @@
...
+        feat_index.resize(n);
+      }
+      {
+        /* determine layout of data */
+        const auto nrow = info.num_row;
+        const auto ncol = info.num_col;
+        const auto nnz = info.num_nonzero;
+        // number of discrete bins for feature 0
+        const unsigned nbins_f0 = gmat.cut->row_ptr[1] - gmat.cut->row_ptr[0];
+        if (nrow*ncol == nnz) {
```
Comment (tqchen) [inline src/tree/updater_fast_hist.cc]: space between operands in multiplication

File: src/tree/updater_fast_hist.cc [hunk 6]
```diff
@@ -0,0 +1,702 @@
...
+          // dense data with zero-based indexing
+          data_layout_ = kDenseDataZeroBased;
+        } else if (nbins_f0 == 0 && nrow*(ncol-1) == nnz) {
```
Comment (tqchen) [inline src/tree/updater_fast_hist.cc]: space between operands

File: src/tree/updater_fast_hist.cc [hunk 7]
```diff
@@ -0,0 +1,702 @@
...
+          // dense data with one-based indexing
+          data_layout_ = kDenseDataOneBased;
+        } else {
+          // sparse data
+          data_layout_ = kSparseData;
+        }
+      }
+      if (data_layout_ == kDenseDataZeroBased || data_layout_ == kDenseDataOneBased) {
+        const std::vector<unsigned>& row_ptr = gmat.cut->row_ptr;
+        const size_t nfeature = row_ptr.size()-1;
```
Comment (tqchen) [inline src/tree/updater_fast_hist.cc]: space between operands

File: src/tree/updater_fast_hist.cc [hunk 8]
```diff
@@ -0,0 +1,702 @@
...
+        size_t min_nbins_per_feature = 0;
+        for (size_t i = 0; i < nfeature; ++i) {
+          const unsigned nbins = row_ptr[i+1] - row_ptr[i];
```
Comment (tqchen) [inline src/tree/updater_fast_hist.cc]: space between operands

File: src/tree/updater_fast_hist.cc [hunk 9]
```diff
@@ -0,0 +1,702 @@
...
+      bst_uint split_cond = 0;
+      // set the condition
+      for (unsigned i = gmat.cut->row_ptr[fid]; i < gmat.cut->row_ptr[fid + 1]; ++i) {
+        if (split_pt == gmat.cut->cut[i]) split_cond = i;
+      }
+
+      const auto& rowset = row_set_collection_[nid];
+      if (data_layout_ == kDenseDataZeroBased || data_layout_ == kDenseDataOneBased) {
+        /* specialized code for dense data */
+        const size_t column_offset = (data_layout_ == kDenseDataOneBased) ? (fid-1): fid;
```
Comment (tqchen) [inline src/tree/updater_fast_hist.cc]: space netween operands

File: src/tree/updater_fast_hist.cc [hunk 10]
```diff
@@ -0,0 +1,702 @@
...
+
+    /* tree growing policies */
+    struct ExpandEntry {
+      int nid;
+      int depth;
+      bst_float loss_chg;
+      ExpandEntry(int nid, int depth, bst_float loss_chg)
+        : nid(nid), depth(depth), loss_chg(loss_chg) {}
+    };
+    inline static bool depth_wise(ExpandEntry lhs, ExpandEntry rhs) {
```
Comment (tqchen) [inline src/tree/updater_fast_hist.cc]: The priority queue uses heap, which is in-stable, how do we break the ties? Add a timestamp of insertsion, when depth equals each other, select smaller timestamp

File: src/common/row_set.h (hunk 1 above)
Comment (hcho3) [inline src/common/row_set.h]: done

File: src/tree/updater_fast_hist.cc (hunk 2 above)
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: fixed

File: src/tree/updater_fast_hist.cc (hunk 4 above)
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: fixed

File: src/tree/updater_fast_hist.cc (hunk 5 above)
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: fixed

File: src/tree/updater_fast_hist.cc (hunk 6 above)
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: fixed

File: src/tree/updater_fast_hist.cc (hunk 7 above)
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: fixed

File: src/tree/updater_fast_hist.cc (hunk 8 above)
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: fixed

File: src/tree/updater_fast_hist.cc (hunk 9 above)
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: fixed

File: src/tree/updater_fast_hist.cc (hunk 10 above)
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: I added timestamp to ExpandEntry to support tie breaking

**Comment by khotilov** [comment]
That's a nice re-use of the existing quantile utilities!

However, while giving it a brief try, it seems to be working with sparse data, but produces a trivial model as soon as I convert the data to dense matrices:
```
> data(agaricus.train, package='xgboost')
> data(agaricus.test, package='xgboost')
> 
> dtrain <- xgb.DMatrix(as.matrix(agaricus.train$data), label = agaricus.train$label)
> dtest <- xgb.DMatrix(as.matrix(agaricus.test$data), label = agaricus.test$label)
> watchlist <- list(eval = dtest, train = dtrain)
> 
> param <- list(max_depth = 2, eta = 1, silent = 1, nthread = 2, 
+               objective = "binary:logistic", eval_metric = "auc",
+               tree_method="hist", max_bin = 256, verbose=1)
> bst <- xgb.train(param, dtrain, nrounds = 2, watchlist)
[13:54:53] Tree method is selected to be 'hist', which uses histogram aggregation for faster training. Using default sequence of updaters: grow_fast_histmaker,prune
[13:54:53] amalgamation/../src/tree/updater_fast_hist.cc:56: Generating gmat: 0.0140374 sec
[13:54:54] amalgamation/../src/tree/updater_fast_hist.cc:227: 
InitData:          0.00 (50.00%)
InitNewNode:       0.00 ( 0.00%)
BuildHist:          0.00 (50.00%)
EvaluateSplit:     0.00 ( 0.00%)
ApplySplit:        0.00 ( 0.00%)
========================================
Total:             0.00
[1]	eval-auc:0.500000	train-auc:0.500000 
[13:54:54] amalgamation/../src/tree/updater_fast_hist.cc:227: 
InitData:          0.00 (  nan%)
InitNewNode:       0.00 (  nan%)
BuildHist:          0.00 (  nan%)
EvaluateSplit:     0.00 (  nan%)
ApplySplit:        0.00 (  nan%)
========================================
Total:             0.00
[2]	eval-auc:0.500000	train-auc:0.500000 
> xgb.dump(bst)
[1] "booster[0]"           "0:leaf=-0.0715053"    "booster[1]"           "0:leaf=-7.42913e-005"
```

**Comment by hcho3** [comment]
@khotilov Thanks a lot for the report. We are working hard to resolve this issue.

**Comment by khotilov** [comment]
@hcho3 Thanks! The last commit does fix the problem.

I have a question about the need to pad the cut values by `rt_eps = 1e-6f`. This constant was primarily used to compare loss values; but here you are re-purposing it and are subtracting it from absolute values of features in a model. This makes things dependent on feature scales (which is unusual condition for tree models):

- obviously, one cannot have features that are on small absolute scales comparable to rt_eps
- for features with large values, subtracting rt_eps would make no numerical effect - which is like having no rt_eps at all.

What was the reason for this epsilon-padding? And if it is really needed, would it make sense to make it scale-adaptive?

**Comment by hcho3** [comment]
@khotilov You're right in that padding would be problematic for "small" continuous features. We removed the padding from the code.
The original reason for the padding was to make sure that categorial/ordinal feature values fall into correct bins. We created a special code to deal with this scenario.

**Comment by khotilov** [comment]
To clarify: while the rt_eps is also used in colmaker in + or - combination with feature values (not just for loss change, as I remembered it earlier), it is used in there in such ways that it still works with small-valued features that I was too lazy to transform. But things break in fast_histmaker when bins are padded by rt_eps that is relatively large to feature scale.

And here's one more edge case that breaks fast_histmaker:
```R
library(data.table)
library(xgboost)

# simulate binary outcome dependent on one feature
N <- 10000
Scale <- 1 #0.00000001
set.seed(1)
#X <- data.table(x = 1.*rbinom(N,1,0.05))
X <- data.table(x = sample(-1:1, N, replace = T, prob = c(0.2,0.6,0.2))  )
pr = X[, plogis(-3 + 2.5 * x^2)]
y = rbinom(N, 1, pr) 
X[, x := Scale * x]
plot(factor(y) ~ factor(X$x))
#X[sample.int(N,100), x:=NA]

dtrain <- xgb.DMatrix(as.matrix(X[itr]), label=y[itr])

param <- list(objective="binary:logistic", eval_metric="auc", nthread=4, 
              max_depth=1, eta = 0.5, base_score=mean(getinfo(dtrain, "label")),
              tree_method="hist", max_bin = 2, verbose = 1)
bst <- xgb.train(param, dtrain, 5)
```
this results in
image
```
[20:12:50] Tree method is selected to be 'hist', which uses histogram aggregation for faster training. Using default sequence of updaters: grow_fast_histmaker,prune
[20:12:50] amalgamation/../src/tree/../common/quantile.h:384:  check quantile stats, nbig=0, n=1
[20:12:50] amalgamation/../src/tree/../common/quantile.h:385:  srcsize=3, maxsize=2, range=3050, chunk=6100
[20:12:50] amalgamation/../src/tree/../common/quantile.h:284: [0] rmin=0, rmax=1049, wmin=1049, v=-1
[20:12:50] amalgamation/../src/tree/../common/quantile.h:284: [1] rmin=1049, rmax=4099, wmin=3050, v=0
[20:12:50] amalgamation/../src/tree/../common/quantile.h:284: [2] rmin=4099, rmax=5000, wmin=901, v=1
Error in xgb.iter.update(bst$handle, dtrain, iteration - 1, obj) : 
  [20:12:50] amalgamation/../src/tree/../common/quantile.h:388: Check failed: nbig < n - 1 quantile: too many large chunk
```

**Comment by khotilov** [comment]
@hcho3 I've posted my last comment before I saw yours. The last updates makes it possible to use small-scale features. Thanks! 

The `max_bins=2` error is still there though.

**Comment by tqchen** [comment]
@khotilov It appears that you hit a corner case of quantile sketch. 

Can you see if https://github.com/dmlc/xgboost/pull/1959 fixes the problem?

**Comment by khotilov** [comment]
@tqchen Thanks, that does fix it.

**Comment by hcho3** [comment]
@tqchen @khotilov #1959 indeed fixes the problem with `max_bin=2`. I've incorporated the fix into this pull request. I've also added a small python test to test for `max_bin=2`.

**Comment by CodingCat** [comment]
so, is it safe to say it is ready to add language-bindings?

**Comment by tqchen** [comment]
yap, I think these options should be exposed.

**Pull request merged**